source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+// Note the Quic plugin strategy relies on link-time binding rather than inheritance.
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length) {
+    size_t total_length = sizeof(Envoy::Buffer::BufferFragmentImpl) + length;
+    char* mem = allocator->New(total_length);
+    auto fragment = new (mem + length) Envoy::Buffer::BufferFragmentImpl(
+        mem, length,
+        [allocator](const void* data, size_t, const Envoy::Buffer::BufferFragmentImpl*) {
+          allocator->Delete(const_cast<char*>(static_cast<const char*>(data)));
+        });
+    single_slice_buffer_.addBufferFragment(*fragment);
+  }
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance with first |length| bytes in it.
+  // Data will be moved from |buffer| to this mem slice.
+  // Prerequisite: |buffer| has at least |length| bytes of data and not empty.
+  explicit QuicMemSliceImpl(Envoy::Buffer::Instance& buffer, size_t length) {
+#ifndef NDEBUG
+    Envoy::Buffer::RawSlice slice;
+    ASSERT(buffer.getRawSlices(&slice, 1) != 0);
+    ASSERT(slice.len_ == length);
+#endif
+    single_slice_buffer_.move(buffer, length);
+    ASSERT(single_slice_buffer_.getRawSlices(nullptr, 0) == 1);
+  }
+
+  QuicMemSliceImpl(const QuicMemSliceImpl& other) = delete;
+  QuicMemSliceImpl& operator=(const QuicMemSliceImpl& other) = delete;
+
+  // Move constructors. |other| will not hold a reference to the data buffer
+  // after this call completes.
+  QuicMemSliceImpl(QuicMemSliceImpl&& other) {
```

Comment:
Is other.Reset() necessary here?

Can you call move assignment from here instead?

QuicMemSliceImpl(QuicMemSliceImpl&& other) {
*this = std::move(other);
}

Reply:
> std::move() doesn't move data from Buffer::Instance automatically.

Not automatically, but it will cause the move assignment function(implemented below) to be called, inside that function, the Buffer::Instance is moved.

> Reset is not needed actually.

Reply:
done